cmd/kubeadm/app/cmd/images.go

```diff
@@ -0,0 +1,112 @@
+func NewCmdImages(out io.Writer) *cobra.Command {
+	cfg := &kubeadmapiext.MasterConfiguration{}
+	legacyscheme.Scheme.Default(cfg)
+	var cfgPath string
+
+	cmd := &cobra.Command{
+		Use:   "images",
+		Short: "Print a list of images kubeadm will use. The configuration file is used in case any images or image repositories are customized.",
+		Run: func(_ *cobra.Command, _ []string) {
+			internalcfg := &kubeadmapi.MasterConfiguration{}
+			legacyscheme.Scheme.Convert(cfg, internalcfg, nil)
+			i, err := NewImages(cfgPath, internalcfg)
+			kubeadmutil.CheckErr(err)
+			kubeadmutil.CheckErr(i.Run(out))
+		},
+	}
+	AddImagesFlags(cmd.PersistentFlags(), &cfgPath)
+
+	return cmd
+}
+
+// NewImages returns a "kubeadm images" command
+func NewImages(cfgPath string, cfg *kubeadmapi.MasterConfiguration) (*Images, error) {
+	if cfgPath != "" {
+		b, err := ioutil.ReadFile(cfgPath)
+		if err != nil {
+			return nil, fmt.Errorf("unable to read config from %q [%v]", cfgPath, err)
+		}
+		if err := kuberuntime.DecodeInto(legacyscheme.Codecs.UniversalDecoder(), b, cfg); err != nil {
+			return nil, fmt.Errorf("unable to decode config from %q [%v]", cfgPath, err)
+		}
+	}
+
+	return &Images{
+		cfg: cfg,
+	}, nil
+}
+
+// Images defines the struct used for "kubeadm images"
+type Images struct {
+	cfg *kubeadmapi.MasterConfiguration
+}
+
+// Run runs the images command and writes the result to the io.Writer passed in
+func (i *Images) Run(out io.Writer) error {
+	imgs := []string{}
+	imgs = append(imgs, images.GetCoreImage(constants.KubeAPIServer, i.cfg.ImageRepository, i.cfg.KubernetesVersion, i.cfg.UnifiedControlPlaneImage))
+	imgs = append(imgs, images.GetCoreImage(constants.KubeControllerManager, i.cfg.ImageRepository, i.cfg.KubernetesVersion, i.cfg.UnifiedControlPlaneImage))
+	imgs = append(imgs, images.GetCoreImage(constants.KubeScheduler, i.cfg.ImageRepository, i.cfg.KubernetesVersion, i.cfg.UnifiedControlPlaneImage))
+	imgs = append(imgs, fmt.Sprintf("%v/%v-%v:%v", i.cfg.ImageRepository, constants.KubeProxy, runtime.GOARCH, kubeadmutil.KubernetesVersionToImageTag(i.cfg.KubernetesVersion)))
+	imgs = append(imgs, fmt.Sprintf("%v/pause-%v:%v", i.cfg.ImageRepository, runtime.GOARCH, "3.1"))
+
+	// if etcd is not external then add the image as it will be required
+	if len(i.cfg.Etcd.Endpoints) == 0 {
+		imgs = append(imgs, images.GetCoreImage(constants.Etcd, i.cfg.ImageRepository, i.cfg.KubernetesVersion, i.cfg.Etcd.Image))
+	}
+
+	dnsImage := fmt.Sprintf("%v/k8s-dns-kube-dns-%v:%v", i.cfg.ImageRepository, runtime.GOARCH, dns.GetDNSVersion(nil, constants.KubeDNS))
+	if features.Enabled(i.cfg.FeatureGates, features.CoreDNS) {
+		dnsImage = fmt.Sprintf("coredns/coredns:%v", dns.GetDNSVersion(nil, constants.CoreDNS))
+	}
+	imgs = append(imgs, dnsImage)
+
+	for _, img := range imgs {
+		fmt.Fprintln(out, img)
+	}
+
+	return nil
+}
+
+// AddImagesFlags adds the flags required for "kubeadm images"
+func AddImagesFlags(flagSet *flag.FlagSet, cfgPath *string) {
```

Comment:
Personally I'm in favour of providing a simpler UX (less flags) than `kubeadm init`/`kubeadm phases`, but I'm a little bit concerned if the users will find the lack of flags confusing when invoking the command without a config file.
@chuckha   What about making the `--config` flag mandatory? (So there is only a supported path fully consistent with `kubeadm init`/`kubeadm phases`)

Reply:
@fabriziopandini is a config file required for `kubeadm init` if no flags are provided? I didn't know that. I agree the UX should be as simple as possible. I think I agree here if `kubeadm init` without a config or flags means nothing then `kubeadm config list-images` should mean nothing as well.

Reply:
Someone brought this up the other day, and it is definitely tangential to this PR, but we really do need to have a command to generate a config populated with the default values...  

Now back on topic, I think requiring a config file is a bit of a poor experience, especially since we should have a happy path to an easy default install. I also think that it makes sense to have a handful of command line flags that can be used for common and trivial config options.

Reply:
Using the existing set of options exposed in v1.10.2:

```sh
vagrant@ubuntu:~$ kubeadm init --help          
Run this command in order to set up the Kubernetes master.                                    

Usage:                                         
  kubeadm init [flags]                         

Flags:                                         
      --apiserver-advertise-address string   The IP address the API Server will advertise it's listening on. Specify '0.0.0.0' to use the address of the default network interface.          
      --apiserver-bind-port int32            Port for the API Server to bind to. (default 6443)                                                                                              
      --apiserver-cert-extra-sans strings    Optional extra Subject Alternative Names (SANs) to use for the API Server serving certificate. Can be both IP addresses and DNS names.          
      --cert-dir string                      The path where to save and store the certificates. (default "/etc/kubernetes/pki")                                                              
      --config string                        Path to kubeadm config file. WARNING: Usage of a configuration file is experimental.                                                            
      --cri-socket string                    Specify the CRI socket to connect to. (default "/var/run/dockershim.sock")                                                                      
      --dry-run                              Don't apply any changes; just output what would be done.                                                                                        
      --feature-gates string                 A set of key=value pairs that describe feature gates for various features. Options are:                                                         
Auditing=true|false (ALPHA - default=false)    
CoreDNS=true|false (BETA - default=false)      
DynamicKubeletConfig=true|false (ALPHA - default=false)                                       
SelfHosting=true|false (ALPHA - default=false) 
StoreCertsInSecrets=true|false (ALPHA - default=false)                                        
  -h, --help                                 help for init                                    
      --ignore-preflight-errors strings      A list of checks whose errors will be shown as warnings. Example: 'IsPrivilegedUser,Swap'. Value 'all' ignores errors from all checks.          
      --kubernetes-version string            Choose a specific Kubernetes version for the control plane. (default "stable-1.10")                                                             
      --node-name string                     Specify the node name.                           
      --pod-network-cidr string              Specify range of IP addresses for the pod network. If set, the control plane will automatically allocate CIDRs for every node.                        --service-cidr string                  Use alternative range of IP address for service VIPs. (default "10.96.0.0/12")                                                                  
      --service-dns-domain string            Use alternative domain for services, e.g. "myorg.internal". (default "cluster.local")                                                           
      --skip-token-print                     Skip printing of the default bootstrap token generated by 'kubeadm init'.                                                                       
      --token string                         The token to use for establishing bidirectional trust between nodes and masters.                                                                
      --token-ttl duration                   The duration before the bootstrap token is automatically deleted. If set to '0', the token will never expire. (default 24h0m0s)   
```
There appear to be a few different "classes" of options that we expose today.
* command options that do not make sense in the config file:
  * --config
  * -h, --help
  * --dry-run
  * --ignore-preflight-errors
  * --skip-token-print
* simple, common config overrides
  * --kubernetes-version
  * --node-name
  * --pod-network-cidr
  * --service-cidr
  * --cri-socket
* feature gates
* more complex config options, or less common overrides that could have wider implications
  * --apiserver-advertise-address
  * --apiserver-bind-port
  * --apiserver-cert-extra-sans
  * --cert-dir
  * --service-dns-domain

I think the first two "classes" definitely make sense to bubble up as command line options. For less common config options or where common configuration overrides would require multiple command line options to be specified (as in the case of overriding the api server advertise address and port combination) make more sense in the config file.

For feature gates, I think that is a bigger discussion that needs to happen. I think we've been a bit too cavalier with our use of feature gates to date. I don't see an issue with bubbling feature gates up to the command line options, but I think we need to be more judicious in what becomes a feature gate vs a new configuration option.

Reply:
So luckily for this PR the only one of those that will affect the output here are

`--kubernetes-version` and `feature-gates=CoreDNS=true`.

I can add the kubernetes version and then in a following PR I will want to refactor the config subcommand to be a package.